File: 92-reverse-linked-list-ii/reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* reverseList(ListNode* head,int count)
    {
        ListNode *curr=head,*prev=NULL;
        while(curr && count)
        {
           ListNode *Next=curr->next;
           curr->next=prev;
           prev=curr;
           curr=Next; 
           count--;
        }

        // making connection with the remaining list in the right side.
        head->next=curr;
        return prev;
    }

    ListNode* reverseBetween(ListNode* head, int left, int right) 
    {
        if(left==right)
        return head;

        ListNode* dummy=new ListNode(0);
        dummy->next=head;
        ListNode *leftEnd=dummy;

        for(int steps=0;steps<left-1;steps++)
        leftEnd=leftEnd->next;

        ListNode* reversedHead=reverseList(leftEnd->next,right-left+1);
        leftEnd->next=reversedHead;
        return dummy->next;  
    }
};

File: 92-reverse-linked-list-ii/reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode reverse(ListNode head, int steps){
        ListNode prev = null;
        ListNode curr = head;
        while(steps-- > 0){
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        // make connection
        head.next = curr;
        return prev;
    }
    
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummyHead = new ListNode(-1);
        dummyHead.next = head;
        ListNode curr = dummyHead;
        for(int i = 1; i <= left - 1; i++){
            curr = curr.next;
        }
        curr.next = reverse(curr.next, right - left + 1);
        return dummyHead.next;
    }
}